### PR TITLE
Slot marshaling

### PR DESCRIPTION
**Pull request by SrsBusiness**
1. cNBT is now a submodule
2. Functions marshal_slot() and demarshal_slot() now handle slot data, with the help of cNBT. These have been incorporated into format_packet and decode_packet, so we should be able to handle all slot packets now.

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[submodule "cNBT"]
+	path = cNBT
+	url = https://github.com/FliPPeh/cNBT.git
```
Comment (JDongian) [inline .gitmodules]: perhaps change this to our fork, for stability reasons

File: Makefile [hunk 2]
```diff
@@ -3,50 +3,65 @@ VERSION = 0.0.1
 CC      = clang
 SA      = scan-build
 
-DIR 	:= $(pwd)
+DIR 	:= $(shell pwd)
```
Comment (JDongian) [inline Makefile]: Could you inform me on what this does?

File: src/marshal.c [hunk 3]
```diff
@@ -92,13 +96,14 @@ uint64_t reverse(uint64_t number, int len)
...
         }
     }
-    //printf("Reversed: %llx, size: %d\n", number, len);
     return number;
 }
 
 size_t format_sizeof(char c)
 {
     switch(c) {
+    case 'k':
```
Comment (JDongian) [inline src/marshal.c]: I'll need to add this to the documentation.

File: src/marshal.c [hunk 4]
```diff
@@ -168,6 +173,11 @@ int format_packet(bot_t *bot, void *_packet_data, void *_packet_raw)
         size = format_sizeof(*fmt);
         packet_data = (char *)align(packet_data, size);
         switch(*fmt) {
+        case 'k': { /* k is for slot. What, do you have a better idea? */
```
Comment (JDongian) [inline src/marshal.c]: I would have used `S`, but whatever

File: Makefile (hunk 2 above)
Comment (SrsBusiness) [inline Makefile]: shell is a keyword in make that runs the argument in your shell. That line is basically command substitution in shell languages.

Comment (SrsBusiness) [inline Makefile]: I actually don't remember why I put that there since we never use it

File: src/marshal.c [hunk 5]
```diff
@@ -227,13 +237,98 @@ int format_packet(bot_t *bot, void *_packet_data, void *_packet_raw)
...
+    /* Read 2 bytes (damage) */
+    memcpy(&slot_data->damage, data, sizeof(slot_data->damage));
+    data += sizeof(slot_data->damage);
+    slot_data->damage = ntohs(slot_data->damage);
+
+    uint64_t nbt_length;
+    if(*data) { /* If next byte is non-zero */
+        nbt_node *tree = nbt_parse(data, space_left);
+        if (!tree) {/* Error */
+            exit(1);
```
Comment (JDongian) [inline src/marshal.c]: // TODO: add error handling

File: src/marshal.c [hunk 6]
```diff
@@ -227,13 +237,98 @@ int format_packet(bot_t *bot, void *_packet_data, void *_packet_raw)
     }
 
     varlen = varint32_encode(packet_raw - save, varint, 5);
-    if(packet_raw - save + varlen > len)
+    if(packet_raw - save + varlen > len) {
```
Comment (JDongian) [inline src/marshal.c]: I want the length checking style to match `decode_packet`.

File: src/marshal.c [hunk 7]
```diff
@@ -254,29 +350,53 @@ int decode_packet(bot_t *bot, void *_packet_raw, void *_packet_data)
...
 
     uint64_t host_order;
     while(*fmt) {
         size = format_sizeof(*fmt);
         packet_data = (char *)align(packet_data, size);
         switch(*fmt) {
+        case 'k':
+            /* Notch decided it was a good idea not to tell us the length of
+             * the nbt binary data. cNBT expects us the know this length. We
+             * use the number of bytes left in the packet.
```
Comment (JDongian) [inline src/marshal.c]: Not sure that using remaining bytes will always work, especially with compression. We'll fix it when it breaks.

Comment (JDongian) [inline src/marshal.c]: Does this work even for http://wiki.vg/Protocol#Player_Block_Placement ?

File: src/marshal.c [hunk 8]
```diff
@@ -242,6 +337,7 @@ int decode_packet(bot_t *bot, void *_packet_raw, void *_packet_data)
     char *packet_raw = _packet_raw;
     char *packet_data = _packet_data;
 
+    uint64_t bytes_read = 0; /* bytes read from packet_raw */
```
Comment (JDongian) [inline src/marshal.c]: What's the point of this variable? Would it be better to just do `packet_raw - _packet_raw`?

File: test/packet_test.c [hunk 9]
```diff
@@ -18,7 +18,8 @@ int random_fmt(char *, int);
 FILE *urandom;
 
 
-void hexdump(char *title, void *buffer, int len) {
+void hexdump(char *title, void *buffer, int len)
```
Comment (JDongian) [inline test/packet_test.c]: why hasn't this been deleted?

**Comment by JDongian** [comment]
Looks great overall. However, there are some small details I want to work out. You probably want to rebase a little bit, as well.

File: src/marshal.c [hunk 10]
```diff
@@ -227,13 +237,98 @@ int format_packet(bot_t *bot, void *_packet_data, void *_packet_raw)
...
+
+    uint64_t nbt_length;
+    if(*data) { /* If next byte is non-zero */
+        nbt_node *tree = nbt_parse(data, space_left);
+        if (!tree) {/* Error */
+            exit(1);
+        }
+        slot_data->tree = tree;
+        struct buffer nbt_bin = nbt_dump_binary(tree);
+        nbt_length = nbt_bin.len;
```
Comment (JDongian) [inline src/marshal.c]: Why is `nbt_length` not equal to `space_left`, as we have previously assumed?

Comment (SrsBusiness) [inline src/marshal.c]: Space_left is not the length of the nbt data. It is the number of free bytes left in packet_data for the nbt data. This is to ensure that nbt_parse won't overflow packet_data.

File: test/packet_test.c (hunk 9 above)
Comment (SrsBusiness) [inline test/packet_test.c]: I'll remove it

File: src/marshal.c (hunk 8 above)
Comment (SrsBusiness) [inline src/marshal.c]: Yeah probably. I'll change that and see if it works.

File: src/marshal.c (hunk 7 above)
Comment (SrsBusiness) [inline src/marshal.c]: See comment below about what the parameter space_left represents. Actually the issue of dealing with incoming NBT data without knowing the length is up for discussion.

File: src/marshal.c (hunk 6 above)
Comment (SrsBusiness) [inline src/marshal.c]: Yeah I'm really sick of this boilerplate bounds checking code every time we add to the buffer. I think we should consider using a buffer struct, possibly even the one that cNBT has (there's already code in mcc that uses it, see line 279 and 322). That saves us passing the length around and we can modify push() to use it.

Comment (JDongian) [inline src/marshal.c]: +1, but sounds like enough for a separate, smaller, pull request

File: src/marshal.c (hunk 10 above)
Comment (JDongian) [inline src/marshal.c]: if it works, then w/e. I'll move this to a new issue.

**Comment by JDongian** [comment]
dead

**Pull request closed without merging**